datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -189,36 +182,59 @@ public void start(CheckpointProvider checkpointProvider) {
+   * If any of them are not running or not healthy, Restart them.
    */
-  protected void restartIfNotRunning(DatastreamTask task) {
-    if (!isTaskRunning(task)) {
-      _logger.warn("Detected that the kafka connector task is not running for datastream task {}. Restarting it", task);
-      boolean stopped = stopTask(task);
-      if (stopped) {
-        createKafkaConnectorTask(task);
-      } else {
-        _logger.error("Datastream task {} could not be stopped.", task);
+  protected void restartTasksIfNotRunning() {
+    synchronized (_runningTasks) {
+      if (_runningTasks.isEmpty()) {
+        _logger.info("connector received no datastreams tasks yet.");
+        return;
       }
+
+      _logger.info("Checking the status of {} running connector tasks.", _runningTasks.size());
+
+      Set<DatastreamTask> deadDatastreamTasks = new HashSet<>();
+      _runningTasks.keySet().forEach((datastreamTask) -> {
+        if (!isTaskRunning(datastreamTask)) {
+          _logger.warn("Detected that the kafka connector task is not running for datastream task {}. Restarting it",
+              datastreamTask);
+          boolean stopped = stopTask(datastreamTask);
+          if (stopped) {
+            deadDatastreamTasks.add(datastreamTask);
+          } else {
+            _logger.error("Datastream task {} could not be stopped.", datastreamTask);
+          }
+        } else {
+          _logger.info("Connector task for datastream task {} is healthy", datastreamTask);
+        }
+      });
+
+      deadDatastreamTasks.forEach(datastreamTask -> {
+        _logger.warn("Creating a new connector task for the datastream task {}", datastreamTask);
+        _runningTasks.remove(datastreamTask);
+        createKafkaConnectorTask(datastreamTask);
+      });
     }
   }
 
   /**
    * Stop the datastream task and wait for it to stop. If it has not stopped within a timeout, interrupt the thread.
+   * The _runningTasks must be locked before calling this method
+   * @param datastreamTask Datastream task that needs to stopped.
+   * @return true if it was successfully stopped, false if it was not.
    */
   private boolean stopTask(DatastreamTask datastreamTask) {
     try {
-      AbstractKafkaBasedConnectorTask kafkaTask = _runningTasks.get(datastreamTask);
+      ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(datastreamTask);
+      AbstractKafkaBasedConnectorTask kafkaTask = connectorTaskEntry.getConnectorTask();
       kafkaTask.stop();
       boolean stopped = kafkaTask.awaitStop(CANCEL_TASK_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
       if (!stopped) {
         _logger.warn("Task {} took longer than {} ms to stop. Interrupting the thread.", datastreamTask,
             CANCEL_TASK_TIMEOUT.toMillis());
-        _taskThreads.get(datastreamTask).interrupt();
+        connectorTaskEntry.getThread().interrupt();
       }
-      _runningTasks.remove(datastreamTask);
-      _taskThreads.remove(datastreamTask);
       return true;
```

Comment:
let's discuss this in person - basically what are the repercussions if we race with when the thread gets interrupted and the countdown latch reflects the fact,  and when we check the status? What if it does get interrupted and we miss this?

Reply:
Good point. The interrupted thread is most certainly not going to react so quickly.

Perhaps we can `return awaitStop()` after interruption (prolly with a shorter timeout than `CANCEL_TASK_TIMEOUT` though), or maybe we can return `false` after interruption so that `restartDeadTasks()` keeps the task around until the next health-check. Bottom line is: I like the idea of seeing a flow of warning logs for rogue tasks until they comply. 

Two important questions beg themselves though:
1. Is it okay if we keep trying to stop a rogue task first before spawning a new one? 
2. Conversely, is it okay to spawn a new task in the presence of an incompliant rogue task?

Let's discuss this in person, I agree.

Reply:
We have decided to awaitStop() for a shorter period of time after interrupting, and adding a warning log indicating that the task may not have stopped. In the future if we see these often, we will add a metric to track this.